include/h2o/stats.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2021 Fastly
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#ifndef h2o__stats_h
+#define h2o__stats_h
+
+#include <stdint.h>
+#include <assert.h>
+#include <string.h>
+#include <stddef.h>
+#include <stdio.h>
+
+struct st_h2o_context_t;
+#define H2O_STATS_MAX_BUCKETS 100
+
+typedef struct st_h2o_stat_ops_t {
+    h2o_iovec_t name;
+    h2o_iovec_t desc;
+    size_t offset;
+    size_t nr_elements;
+    const char **labels;
+    void (*aggregate)(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, struct st_h2o_context_t *per_thread);
+    void (*stringify)(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
+    void (*prometheus)(const struct st_h2o_stat_ops_t *self, struct st_h2o_context_t *agg, h2o_buffer_t **buf);
```

Comment:
This is related to my other comment, but I think we should either tie stat_ops to a particular struct that contains a set of known variables, or make it be used against any struct.

If we are going to bind this functionality to `h2o_context_t` or (specific type used within `h2o_context_t`), then we'd better move the definition into h2o.h (because it'd be about dealing with types defined there).

OTOH, if we are to add support for collecting stats at each handler, then the latter would be the way to go, meaning that we'd be using `void *` instead of `h2o_context_t *` as a reference to the containers of the values.